test/logflare/teams_test.exs

```diff
@@ -1,44 +1,91 @@
-      assert Teams.get_home_team(team_user).id == home_team.id
-    end
+  test "list_teams/0 fetches data correctly" do
+    teams = insert_list(2, :team)
+    expected = teams |> Enum.map(& &1.id) |> Enum.sort()
+    result = Teams.list_teams() |> Enum.map(& &1.id) |> Enum.sort()
+
+    assert expected == result
+  end
+
+  test "get_team/1 fetches data correctly" do
+    %{id: team_id} = insert(:team)
+
+    assert %Team{id: ^team_id} = Teams.get_team!(team_id)
+  end
+
+  test "get_team_by/1 fetches data correctly" do
+    %{id: team_id, name: name} = insert(:team)
+
+    assert %Team{id: ^team_id} = Teams.get_team_by(name: name)
+  end
+
+  test "preload_user/1 preloads user" do
+    %{id: user_id} = user = insert(:user)
+    team = insert(:team, user: user)
+
+    assert %{user: %User{id: ^user_id}} = Teams.preload_user(team)
+  end
+
+  test "preload_user/1 preloads preload_team_users" do
+    team = insert(:team, user: insert(:user))
+    %{email: user_email} = user = insert(:user)
+    insert(:team_user, team: team, email: user.email)
 
-    test "create_team/2, update_team/2, delete_team/1" do
-      user = insert(:user)
-      assert {:ok, %Team{} = team} = Teams.create_team(user, @valid_attrs)
-      assert team.name == @valid_attrs.name
+    assert %{team_users: [%{email: ^user_email}]} = Teams.preload_team_users(team)
+  end
+
+  test "get_home_team/1 returns a paying user's home team" do
+    # owned by different user
+    home_team = insert(:team)
+    user = insert(:user, team: home_team)
+    team_user = insert(:team_user, email: user.email)
+
+    assert Teams.get_home_team(team_user).id == home_team.id
+  end
 
-      assert {:ok, %Team{} = team} = Teams.update_team(team, @update_attrs)
-      assert team.name == @update_attrs.name
+  test "create_team/2 creates a new team in the database" do
+    name = TestUtils.random_string()
+    user = insert(:user)
+    assert {:ok, %Team{name: ^name}} = Teams.create_team(user, %{name: name})
+  end
+
+  test "update_team/2 updates an existing team" do
+    name = TestUtils.random_string()
+    %{id: id} = team = insert(:team)
+    assert {:ok, %{id: ^id, name: ^name}} = Teams.update_team(team, %{name: name})
+  end
 
-      assert {:ok, %Team{} = team} = Teams.delete_team(team)
+  test "delete_team/1 deletes an existing teams" do
+    %{id: id} = team = insert(:team)
+    assert {:ok, %{id: ^id}} = Teams.delete_team(team)
 
-      assert_raise Ecto.StaleEntryError, fn ->
-        Teams.delete_team(team)
-      end
+    assert_raise Ecto.StaleEntryError, fn ->
+      Teams.delete_team(team)
     end
   end
+
+  test "list_for_user/1 lists all teams of a given user" do
+    user = insert(:user)
+    teams = insert_list(2, :team_user, email: user.email)
+
+    expected = Enum.map(teams, & &1.team.id) |> Enum.sort()
+    result = Enum.map(Teams.list_for_user(user), & &1.id) |> Enum.sort()
+
+    assert expected == result
+  end
+
+  test "get/2 gets a team for a given user and token" do
```

Comment:
should combine this function with `Teams.get_team/1` and pattern match on 2 arity

Reply:
at the moment they have different purposes. `get_team/1` fetches the Team entity, `get/2` fetches the Team or a Team of a Team User the given user might be part of (that was confusing to write...)

but in the end, it has different purposes ... it might need a different name but it's a tricky one to handle

Reply:
seems like you're trying to do too much in one function, and also the function name doesn't self document and convey what is being fetched or checked for.

would suggest having a `Teams.has_team_access?/1` which accepts a User/TeamUser/email and does the check before actually fetching the team with `Teams.get_team/1`

Reply:
that would mean extra queries right? it's a weird balance because I want to fetch all the teams that user has in their dashboard which is the one they own and the ones they have been invited to... honestly I feel that the other methods are the ones wrongly named since they should be "owner" focused fetches 😅

Reply:
if thats the case then we can reuse the `list_teams_by_user_access` query and add a where filter to a team it to to one input team, that way we reuse the access logic. 

I think that having `Team.get_team/2` where first arg is the User/TeamUser is quite self explanatory that it is checking for whether user has access, though it is also ok to separate it out to two functions `Teams.get_team/1` which just retrieves any team, and `Teams.get_team_by_user_access/2` which then extends the has-access query. 

In general we opt for smaller, faster, and more flexible functions and less code dupe, because we can always use cachex if necessary for such rarely-updated tables.

Reply:
in general we want to keep context functions as flexible as possible and don't want to put duplicate business logic in too many functions, as it makes refactoring or logic adjustments more complicated.

Reply:
I will extract the query from `list_teams_by_user_access` and then use composition to add the extra where clause on `get_team_by_user_access` 👀